es72x/src/main/scala/tech/beshu/ror/es/dlsfls/DocumentFieldReader.scala

```diff
@@ -0,0 +1,231 @@
+import tech.beshu.ror.accesscontrol.domain.DocumentField
+import tech.beshu.ror.es.dlsfls.DocumentFieldDirectoryReader.DocumentFieldDirectorySubReader
+import tech.beshu.ror.fls.FieldsPolicy
+import ujson._
+
+import scala.collection.JavaConverters._
+import scala.util.Try
+
+private class DocumentFieldReader(reader: LeafReader, fields: NonEmptySet[DocumentField])
+  extends FilterLeafReader(reader) with Logging {
+
+  private val policy = new FieldsPolicy(fields)
+  private val remainingFieldsInfo = {
+    val fInfos = in.getFieldInfos
+    val newfInfos = if (fInfos.asScala.isEmpty) {
+      logger.warn("original fields were empty! This is weird!")
+      fInfos
+    } else {
+      val remainingFields = fInfos.asScala.filter(f => policy.canKeep(f.name)).toSet
+      new FieldInfos(remainingFields.toArray)
+    }
+    logger.debug(s"always allow: ${Constants.FIELDS_ALWAYS_ALLOW.asScala.mkString(",")}")
+    logger.debug(s"original fields were: ${fInfos.asScala.map(_.name).mkString(",")}")
+    logger.debug(s"new fields are: ${newfInfos.asScala.map(_.name).mkString(",")}")
+    newfInfos
+  }
+
+  override def getFieldInfos: FieldInfos = remainingFieldsInfo
+
+  override def getTermVectors(docID: Int): Fields = {
+    val original = in.getTermVectors(docID)
+    new Fields {
+      override def iterator(): JavaIterator[String] = Iterators.filter(original.iterator, (s: String) => policy.canKeep(s))
+      override def terms(field: String): Terms = if (policy.canKeep(field)) original.terms(field) else null
+      override def size(): Int = remainingFieldsInfo.size
+    }
+  }
+
+  override def getNumericDocValues(field: String): NumericDocValues =
+    if (policy.canKeep(field)) in.getNumericDocValues(field) else null
+
+  override def getBinaryDocValues(field: String): BinaryDocValues =
+    if (policy.canKeep(field)) in.getBinaryDocValues(field) else null
+
+  override def getNormValues(field: String): NumericDocValues =
+    if (policy.canKeep(field)) in.getNormValues(field) else null
+
+  override def getSortedDocValues(field: String): SortedDocValues =
+    if (policy.canKeep(field)) in.getSortedDocValues(field) else null
+
+  override def getSortedNumericDocValues(field: String): SortedNumericDocValues =
+    if (policy.canKeep(field)) in.getSortedNumericDocValues(field) else null
+
+  override def getSortedSetDocValues(field: String): SortedSetDocValues =
+    if (policy.canKeep(field)) in.getSortedSetDocValues(field) else null
+
+  override def getPointValues(field: String): PointValues =
+    if (policy.canKeep(field)) in.getPointValues(field) else null
+
+  override def terms(field: String): Terms =
+    if (policy.canKeep(field)) in.terms(field) else null
+
+  override def getMetaData: LeafMetaData = in.getMetaData
+
+  override def getLiveDocs: Bits = in.getLiveDocs
+
+  override def numDocs: Int = in.numDocs
+
+  override def getDelegate: LeafReader = in
+
+  override def document(docID: Int, visitor: StoredFieldVisitor): Unit = {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Method is longer than 50 lines](https://app.codacy.com/manual/scarduzio/elasticsearch-readonlyrest-plugin/pullRequest?prid=4803411)